File: app/src/main/java/com/example/android/newsapp/refactor/mvp/MainPresenter.java
package com.example.android.newsapp.refactor.mvp;

import com.example.android.newsapp.refactor.model.News;
import com.example.android.newsapp.refactor.model.Result;
import com.example.android.newsapp.refactor.mvp.contract.MainContract;
import com.example.android.newsapp.refactor.usecase.DataCallback;
import com.example.android.newsapp.refactor.usecase.UseCaseFactory;

import java.util.List;

public class MainPresenter extends MainContract.Presenter implements DataCallback<List<Result>> {

    private final UseCaseFactory useCaseFactory;

    public MainPresenter(UseCaseFactory useCaseFactory) {
        this.useCaseFactory = useCaseFactory;
    }

    @Override
    public void onTakenView() {
        super.onTakenView();
        useCaseFactory.getNewsUseCase(this).execute();
    }

    @Override
    public void onSuccess(List<Result> data) {
        view.showNews(data);
    }

    @Override
    public void onFailure(String errorMessage, int errorCode) {
        view.showError(errorMessage);
    }
}

File: app/src/main/kotlin/com/example/android/newsapp/mvp/MainPresenter.kt
package com.example.android.newsapp.mvp


import com.example.android.newsapp.model.Result
import com.example.android.newsapp.mvp.contract.MainContract
import com.example.android.newsapp.usecase.UseCaseFactory
import com.example.android.newsapp.usecase.base.DataCallback


class MainPresenter(private val useCaseFactory: UseCaseFactory) : MainContract.Presenter(), DataCallback<List<Result>> {

    override fun onTakenView() {
        super.onTakenView()
        useCaseFactory.getNewsUseCase(this).execute()
    }

    override fun onSuccess(data: List<Result>) {
        view?.showNews(data)
    }

    override fun onFailure(errorMessage: String, errorCode: Int) {
        view?.showError(errorMessage)
    }
}
